qspi: stm32: Enhance the QSPI stm32 driver

lmajewski opened the pull request:
This pull request provides following enhancements to STM32 QSPI driver:
- Generation of the reset pulse for SPI-NOR memory
- Usage of 4IO for read / write (4READ/4PP)
- Support for different QSPI banks - bank1 and bank2 switch supported
- Support for 4B addressing on spi-nor

Not supported:
- The DMA operation (only interrupt mode supported)

Test HW:
- `nucleo_h743zi` with cape board

Test:
- `west build -p always -b nucleo_h743zi ./zephyr/samples/subsys/fs/littlefs`
- `west debug --gdb /home/lukma/.local/opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gdb-no-py`

Commit:
dts: Add DTS information about QUADSPI on stm32h7

Now in Zephyr only stm32f7 processors have by default provided support
for QSPI IP block.

The same IP block is also available in stm32h7 (with different offsets),
so proper DTS description shall be provided as well for it.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
qspi: stm32: Add support for 'flash-id' property in 'stm32-qspi' banks

The 'flash-id' property defines the number of QSPI's dedicated GPIO
bank (and flash ID), which will be used to communicate with flash
memory.

For example, on stm32h7xx it is possible to use 'quadspi_bk1_*' and
'quadspi_bk2_*' set of pins, so one may need to select between them
when required.

By default - pins from 'quadspi_bk1_*' bank are used, so the
'flash-id = <2>;' property, when your use case (e.g. PCB design)
requires it, forces usage of 'quadspi_bk2_*' pins.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
samples: dts: Extend nucleo_h743 DTS to support attached QSPI memory

The nucleo_h743 allows extra capes to extend its functionality.

In our case it has been extended to use 32MiB SPI-NOR memory
connected to QUADSPI dedicated pins from block2.

Moreover, upper 16MiB of SPI-NOR memory has been used to store
littlefs data to test 4B addressing mode (which may cause problems
when NOR-flash is larger than 16 MiB).

Now the littlefs sample can be used to test this extension:
west build -p always -b nucleo_h743zi ./zephyr/samples/subsys/fs/littlefs

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
qspi: stm32: Add support for 'reset-gpios' property in 'stm32-qspi-nor'

Some flash memories connected to QUADSPI IP block on stm32[fh]7 devices
require proper reset pulse before configuration.

This patch adds two new properties - the 'reset-gpios' phandle,
which allows specifying GPIO pin for RESETn pulse and
'reset-gpios-duration', which provides the time (in ms) for reset
duration.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
spi: Provide definition for SPI-NOR flash memory control register

This commit provides define for reading the control register (CR) on
the flash memory.
In that register the information about 4B addressing is stored.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
qspi: stm32: Extend the driver to support 4 byte addressing

After this change the stm32 quadspi driver supports the 4 bytes
addressing if the memory supports it.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
qspi: stm32: Add code to switch flash memory to use 4 SIO pins

For proper operation of QUADSPI IP block, both sides of communication;
NOR flash memory and STM32 controller need to support 4 IO
transmission.

After this change the QSPI stm32 driver is able to program NOR flash
memory to switch itself to use all 4 IO lines (SIO[0123]) to transmit
and receive data.

The QE bit (in SPI-NOR's Status Register) is the non-volatile one, so
setting it is done only once (at first boot of the device).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: stm32: Introduce spi-bus-width property for stm32 QSPI

The spi-bus-width property provides information if the target system
has all SIO[0123] pins connected to NOR flash memory.

It must be equal to 4 to enable QSPI 4 IO operation.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
kconfig: stm32: Select CONFIG_USE_STM32_HAL_MDMA when QSPI is used

The stm32 HAL code for qspi (SHA1: 5c8275071ec1cf160bfe8c18bbd93):
stm32cube/stm32h7xx/drivers/src/stm32h7xx_hal_qspi.c implicitly
depends on the stm32h7xx_hal_mdma.c driver as it uses unconditionally
'HAL_MDMA_Abort_IT()' function.

To avoid build breaks - no matter if QSPI IP block works with DMA
or interrupts - the CONFIG_USE_STM32_HAL_MDMA shall be defined.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

erwango commented on dts/bindings/flash_controller/st,stm32-qspi-nor.yaml:
```diff
@@ -43,3 +43,7 @@ properties:
+    spi-bus-width:
+      type: int
+      required: false
+      description: The width of (Q)SPI bus to which flash memory is connected
```
Let's add some info on the expected/supported values

erwango commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -50,3 +50,13 @@ properties:
+
+    flash-id:
+      type: int
+      description: |
```
required: ?
Also, please consider the use of default

lmajewski replied:
I think that the "default" and "required" properties are not needed. This option shall be used when somebody by mistake connect `quadspi_bk2` pins (which are dedicated to connecting second flash in two flash setup) to first flash.

erwango replied:
Please add `required: false`

lmajewski replied:
Ok

erwango reviewed (changes requested):
Thanks for this great and clean addition.
First quick round of review

lmajewski commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -642,6 +645,29 @@ static int spi_nor_process_bfp(const struct device *dev,
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+static int flash_stm32_qspi_gpio_reset(void)
+{
+	const struct device *reset_dev =
```
Ok, I will follow the used pattern.

lmajewski commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -659,6 +663,97 @@ static int qspi_program_addr_4b(const struct device *dev)
+	}
+
+	ret = qspi_write_enable(dev);
+	if (ret)
```
Ok.

lmajewski commented:
@erwango Do you have any more comments?

erwango commented on samples/subsys/fs/littlefs/boards/nucleo_h743zi.conf:
```diff
@@ -0,0 +1,22 @@
+CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=8192
+CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=8192
+
+CONFIG_FLASH_STM32_QSPI=y
```
Nit: Group the CONFIG_FS_LITTLEFS_foo together

erwango reviewed (changes requested):
Minor changes requested, otherwise looks good.
Haven't been able to test on H7 but no regression on L4

erwango reviewed (approved):
LGTM.

@lmajewski If, interesredt, you can add yourself in CODEOWNER file in front of this PR. This could be done in a follow up PR.

lmajewski commented:
@erwango I will prepare proper patch as a follow up. No problem. Thanks for the review :-)

lmajewski commented:
> LGTM.
> 

Are there any other comments regarding this PR? Is it eligible for being pulled?

erwango commented:
> Is it eligible for being pulled?

@lmajewski, to be eligible for merge, it requires a second approval.
@FRASTM, @ABOSTM , can you review ?

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -20,6 +20,15 @@
+#endif
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
```
The `DT_INST_NODE_HAS_PROP(0, reset_gpios)` is used so often that maybe it deserves some specific define, that would clarify the intention.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -608,6 +680,133 @@ static int setup_pages_layout(const struct device *dev)
+
+	ret = qspi_read_access(dev, &cmd, &reg, sizeof(reg));
+	if (ret == 0 && !(reg & SPI_NOR_4BYTE_BIT)) {
+		return -1;
```
Shouldn't we adapt some errno ERR here? There is clearly intention for this to have some meaning because it gets printed in log here: https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R861

de-nordic replied:
Also in line https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R696 error gets returned from other function, `qspi_send_cmd`, which actually returns negative errno code. So now you have this function either return errno code or hardcoded -1.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +840,66 @@ static int spi_nor_process_bfp(const struct device *dev,
+					data->flag_access_32bit = true;
+					LOG_INF("Flash - address mode: 4B");
+				} else {
+					LOG_ERR("Unable to enter 4B mode: %d\n",
```
I predict the error: it is -1.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -608,6 +680,133 @@ static int setup_pages_layout(const struct device *dev)
+	/* Check if QE bit programming is finished */
+	if (!(reg & SPI_NOR_QE_BIT)) {
+		LOG_ERR("Quad Enable [QE] bit in status reg not set");
+		return -1;
```
Again errno? Fortunately log is, at least, giving us point of failure.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +840,66 @@ static int spi_nor_process_bfp(const struct device *dev,
+	rc = jesd216_bfp_read_support(php, bfp, JESD216_MODE_144, &res);
+	if (IS_ENABLED(STM32_QSPI_QUAD_IO) && rc > 0) {
+		/* Program flash memory to use SIO[0123] */
+		qspi_program_quad_io(dev);
```
The function can fail and report that by returning in band error codes, so shouldn't this be checked here? Before doing anything else below?

lmajewski replied:
+1

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -608,6 +680,133 @@ static int setup_pages_layout(const struct device *dev)
+	cmd.DataMode = QSPI_DATA_1_LINE;
+
+	ret = qspi_read_access(dev, &cmd, &reg, sizeof(reg));
+	if (ret == 0 && !(reg & SPI_NOR_4BYTE_BIT)) {
```
Since you use `if (ret)` in other places to test for non-0 ret, I suppose that this should be `!ret`, the same way you do here https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R754

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +840,66 @@ static int spi_nor_process_bfp(const struct device *dev,
+		/* Program flash memory to use SIO[0123] */
+		qspi_program_quad_io(dev);
+
+		LOG_INF("Mode: 1-4-4 with instr:[0x%x] supported!", res.instr);
```
You can have success info even if the `qspi_program_quad_id` failed, and the `aspi_program_quad_id` has similar log at the end `https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R804` making this line redundant.

lmajewski replied:
There is a difference between those two `LOG_INF()` information.

The one with "Mode: 1-4-4..." informs about the instruction code supported by the flash memory.
The one from the pasted above link informs that in SPI-NOR memory the QSPI mode has been enabled (the status register has correct bit set).

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +840,66 @@ static int spi_nor_process_bfp(const struct device *dev,
+	 */
+	struct jesd216_instr res;
+
+	rc = jesd216_bfp_read_support(php, bfp, JESD216_MODE_144, &res);
```
If `STM32_QSPI_QUAD_IO` is not enabled, the call seems redundant as there is not further processing of information gathered by call, nor the `rc` is used in any way.

lmajewski replied:
+1

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -657,6 +915,13 @@ static int flash_stm32_qspi_init(const struct device *dev)
 	}
 
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	ret = flash_stm32_qspi_gpio_reset(dev);
+	if (ret < 0) {
+		LOG_ERR("QSPI GPIO reset failed (%d)", ret);
+		return ret;
+	}
```
It does not seem that `flash_stm32_qspi_gpio_reset` can ever fail, sot this code is redundant.

lmajewski replied:
+1

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +840,66 @@ static int spi_nor_process_bfp(const struct device *dev,
 	LOG_DBG("Page size %u bytes", data->page_size);
 	LOG_DBG("Flash size %u bytes", flash_size);
+
+	addr_mode = jesd216_bfp_addrbytes(bfp);
+	if (addr_mode == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B) {
+		struct jesd216_bfp_dw16 dw16;
+
+		if (jesd216_bfp_decode_dw16(php, bfp, &dw16) == 0) {
+			/*
+			 * According to JESD216, the bit0 of dw16.enter_4ba
+			 * portion of flash description register 16 indicates
+			 * if it is enough to use 0xB7 instruction without
+			 * write enable to switch to 4 bytes addressing mode.
+			 */
+			if (dw16.enter_4ba & 0x1) {
+				rc = qspi_program_addr_4b(dev);
+				if (rc == 0) {
+					data->flag_access_32bit = true;
+					LOG_INF("Flash - address mode: 4B");
+				} else {
+					LOG_ERR("Unable to enter 4B mode: %d\n",
+						rc);
+					return rc;
+				}
+			}
+		}
+	}
```
Is this still relevant if `STM32_QSPI_QUAD_IO` is not enabled?

Commit:
CODEOWNERS: Add entry for stm32 QSPI driver

I would like to add myself (@lmajewski) as the code owner for
stm32's QSPI driver.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

erwango commented:
@lmajewski I'm setting label DNM to avoid early merge, as I understood you were about to push some of the comments reported by @nvlsianpu. Let me know when that's good for you.

lmajewski commented:
> @lmajewski I'm setting label DNM to avoid early merge, as I understood you were about to push some of the comments reported by @nvlsianpu. Let me know when that's good for you.

Yes, I'm preparing new PR with changes proposed by @de-nordic

erwango commented:
> Yes, I'm preparing new PR with changes proposed by @de-nordic

(I mixed the reviewers names, my bad)
Ok, you could also have fixed current PR, but since that's minor stuff, let's proceed that way.

lmajewski commented:
@erwango @de-nordic 
I've submitted the new PR.

Most notable changes:
- Add CODEOWNERS entry
- Fix for 4READ command's read latency bits (correct naming and number for my flash)
- Fix the QE bit setting (it is a non-volatile bit, so need to be set only once).
- Change the code according to @de-nordic comments

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -100,6 +125,53 @@ static inline void qspi_unlock_thread(const struct device *dev)
+	 * hence it has been assumed that NOR flash memory supporting
+	 * 1-4-4 mode also would support fast page programming.
+	 */
+	if (IS_ENABLED(STM32_QSPI_QUAD_IO) && dev_data->flag_quad_io_en) {
```
I think that the `IS_ENABLED` macro, although will work here, is intended to be used with `CONFIG_` definitions.

lmajewski replied:
I would prefer to leave it here.

de-nordic replied:
The checkpatch is shy to inform you that it would not like that: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

lmajewski replied:
Ok, Then I will use:
`if (STM32_QSPI_QUAD_IO && dev_data->flag_quad_io_en)` directly in the if clause.

erwango commented:
> I've submitted the new PR.

Thanks!
Just a quick clarification (so that we're all clear on what is happening):
What you've done is to push an update of current PR (which is fine and is indeed what was expected).
When you mention "I've submitted the new PR.", this is usually understood as "I have pushed a new branch and a new Github pull request has been created".
So no harm, everything fine, just a small wording issue :-).

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -20,6 +20,17 @@
 
+#if DT_INST_NODE_HAS_PROP(0, spi_bus_width) && \
+	DT_INST_PROP(0, spi_bus_width) == 4
+#define STM32_QSPI_QUAD_IO 1
```
Redefine it as something `_ENABLED` because the `STM32_QSPI_QUAD_IO` sounds like index of some IO or number of IOs.

lmajewski replied:
Maybe it shall be `STM32_QSPI_USE_QUAD_IO` ? 
(The same approach in the same file is with `STM32_QSPI_USE_DMA`)

I would like to avoid the situation when we would have `IS_ENABLED(STM32_QSPI_QUAD_IO_ENABLED)` as it looks
like some kind of the overkill.

de-nordic replied:
Checkpatch failed to detect your usage if `IS_ENABLED` without config, this should generate warning: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

lmajewski replied:
Ok, Then I will use `STM32_QSPI_USE_QUAD_IO` directly in the `if (` statement.

erwango replied:
> Checkpatch failed to detect your usage if `IS_ENABLED` without config, this should generate warning: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

This is not the case anymore, cf https://github.com/zephyrproject-rtos/zephyr/pull/39390

de-nordic replied:
touché

lmajewski replied:
So then I shall use:
`if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) ... )` ?

de-nordic replied:
Yes, you can use it.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -608,6 +682,133 @@ static int setup_pages_layout(const struct device *dev)
+	if (ret) {
+		return ret;
+	}
+	reg |= SPI_NOR_QE_BIT;
+
+	ret = qspi_write_status_register(dev, reg);
```
Just cosmetic: you may want remove empty line at 783 and add one above 782; the bit set is rather prerequisite for the `qspi_write_status_register` than post operation of the `qspi_write_enable`. It is a little bit dangling there.

de-nordic commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -639,9 +842,76 @@ static int spi_nor_process_bfp(const struct device *dev,
+					data->flag_access_32bit = true;
+					LOG_INF("Flash - address mode: 4B");
+				} else {
+					LOG_ERR("Unable to enter 4B mode: %d\n",
+						rc);
```
We can have 100 characters per line now, you do not have to use it everywhere but do not leave the lonely rc dangling in next line, it looks sad.

Commit:
qspi: stm32: Add support for 4 IO read and program (4READ/4PP)

This change provides support for 4 IO read (via 4READ command) and
program (via 4PP). Flash memory pins SIO[0123], CLK and CS are used.

All of them are controlled by stm32 QSPI IP block.

The instruction code for fast reading as well as number of latency
cycles required are read from SFDP structure provided by flash memory.
The number of required read latency cycles when performing reading
is the sum of SPI-NOR memory mode bits and wait states (also named
as 'dummy cycles').

It also has been assumed that memory, which supports fast read (4READ),
also will support fast programming (4PP command) as this information is
not available in SFDP.

One also need to enable the QUAD IO support in board's device tree by
defining 'spi-bus-width = <4>' property. It is required as it may
happen that not all QSPI dedicated pins are used (for example only two
of them are available).

Signed-off-by: Lukasz Majewski <lukma@denx.de>
